### DISCUSSION TITLE
Different configurations that have the same name will not currently work with Monaco

### DISCUSSION THREAD

**Discussion by 163a** [Dynatrace/dynatrace-configuration-as-code#365] [Ideas]
**What is the problem case?**
Currently, Dynatrace allows some configurations to share the same name, even for different configurations. Monaco, however, treats configuration names as a unique key.

This leads to unsupported cases, such as:
- NOT possible to fully backup all configurations JSONs: 
    -   CLI 2.0 download will merge two or more configurations with the same name under the same YAML object. They will reference only one JSON file since Monaco doesn't set a unique JSON, even though Dynatrace has two or more different configurations
    - example with anomaly detection service rules, these `Queue` configs are in-fact two different metric sources.

```
Nodeisdown:
- name: Node is down
Offlinepartitions:
- name: Offline partitions
Poolisdown:
- name: Pool is down
Queuedepthhasdroppedbelowthreshold:
- name: Queue depth has dropped below threshold
- name: Queue depth has dropped below threshold
Queuedepthisreachingmaximum:
- name: Queue depth is reaching maximum
- name: Queue depth is reaching maximum
QueueinhibitsGET:
- name: Queue inhibits GET
- name: Queue inhibits GET
QueueinhibitsPUT:
- name: Queue inhibits PUT
- name: Queue inhibits PUT
Queuemanagerhashighactivechannels:
- name: Queue manager has high active channels
- name: Queue manager has high active channels
Queuemanagerhaslowconnections:
- name: Queue manager has low connections
- name: Queue manager has low connections
Queuemanagerisdown:
- name: Queue manager is down
- name: Queue manager is down
Queuemessagesareaging:
- name: Queue messages are aging
- name: Queue messages are aging
RabbitMQNodefailed:
```

- NOT possible to define two new configs with the same name, since duplicate UID error will be thrown
    - example error
    ```
    2021-07-01 15:49:25 ERROR Validation of pre-prod failed. Found 1 error(s)
    2021-07-01 15:49:25 ERROR 	duplicate UID 'anomaly-detection-metrics/Monaco-check' found in projects/ibmmq/anomaly-detection-metrics/Monaco-check1 and projects/ibmmq/anomaly-detection-metrics/Monaco-check2
    ```
    - example YAML
 ```
config:
- Monaco-check1: OneAgentExtension_AnomalyDetection.json
- Monaco-check2: ActiveGateExtension_AnomalyDetection.json

Monaco-check1:
- name: 'Monaco-check'
Monaco-check2:
- name: 'Monaco-check'

```
 
 **What needs to change?**
 Monaco supporting configurations with the same name, and use something else as the unique ID. Maybe it should be possible to store the existing configurations with the Dynatrace ID? Also throw an error if Monaco detects from API that there are two or more configurations with the same name, so it will prevent overwriting the wrong one until a user selection is made.

**Comment by tatomir146** [comment]
We already had this discussion multiple times and the conclusion was always the same: we are going, at least for now, to stick with name, not IDs, as storing IDs would require additional persistence layer in monaco and we want to keep it as simple as possible. That's also the reason for implementing the duplicated names checker. Even thought Dynatrace sometimes allows you to store different config with same name, what would be the use case for it?

**Reply by 163a** [reply]
Well maybe I didn't make it clear here, but these example configs come from Dynatrace developed extensions. So maybe they need to be corrected by Dynatrace developers firstly for OOTB and turn-key solutions, so it can set the standard for others to follow.
